Streaming.cpp

```diff
@@ -756,139 +750,141 @@ int SoapySidekiq::deactivateStream(SoapySDR::Stream *stream, const int flags,
+                }
+                else
+                {
+                    SoapySDR_logf(SOAPY_SDR_ERROR,
                         "skiq_stop_tx_streaming failed (card %u), status %d",
                         card, status);
-                throw std::runtime_error("");
+                    throw std::runtime_error("");
+                }
             }
         }
-
     }
 
     return 0;
 }
 
-int SoapySidekiq::readStream(SoapySDR::Stream *stream, void *const *buffs,
-                             const size_t numElems, int &flags,
-                             long long &timeNs, const long timeoutUs)
+int SoapySidekiq::readStream(
+    SoapySDR::Stream *stream,
+    void *const *buffs,
+    const size_t numElems,
+    int &flags,
+    long long &timeNs,
+    const long timeoutUs)
 {
-    if (stream != RX_STREAM)
-    {
-        return SOAPY_SDR_NOT_SUPPORTED;
-    }
-    else if (rx_receive_operation_exited_due_to_error)
-    {
-        return SOAPY_SDR_STREAM_ERROR;
-    }
-
-    if (numElems % rx_payload_size_in_words != 0)
-    {
-        SoapySDR_logf(SOAPY_SDR_ERROR, "numElems must be a multiple of the RX block size "
-                     " numElems %d, block size %u",
-                     numElems, rx_payload_size_in_words);
-        throw std::runtime_error("");
-        return 0;
-    }
-
-    long waitTime = timeoutUs;
-
-    // if the user didn't give a waittime then wait a LONG time
-    if (waitTime == 0)
-    {
-        waitTime = SLEEP_1SEC;
-    }
-
-    // see if we have any receive buffers to give
-    while ((rxReadIndex == rxWriteIndex) && (waitTime > 0))
-    {
-        // wait
-        usleep(DEFAULT_SLEEP_US);
-        waitTime -= DEFAULT_SLEEP_US;
-    }
-
-    if (waitTime <= 0)
-    {
-        SoapySDR_log(SOAPY_SDR_DEBUG, "readStream timed out");
-        return SOAPY_SDR_TIMEOUT;
-    }
-
-    char *buff_ptr = (char *)buffs[0];
-    skiq_rx_block_t *block_ptr = p_rx_block[rxReadIndex];
-    char *ringbuffer_ptr = (char *)((char *)block_ptr->data);
-
-    // move to the next buffer in the ring
-    rxReadIndex = (rxReadIndex + 1) % DEFAULT_NUM_BUFFERS;
-
-    if (this->rfTimeSource == true)
-    {
-        timeNs = convert_timestamp_to_nanos(block_ptr->rf_timestamp, rx_sample_rate);
-    }
-    else
-    {
-        timeNs = convert_timestamp_to_nanos(block_ptr->sys_timestamp, sys_freq);
-    }
-
-    flags = SOAPY_SDR_HAS_TIME;
-
-    uint32_t block_num = 0;
-    uint32_t num_blocks = numElems / rx_payload_size_in_words;
+    if (stream != RX_STREAM) return SOAPY_SDR_NOT_SUPPORTED;
+    if (rx_receive_operation_exited_due_to_error) return SOAPY_SDR_STREAM_ERROR;
+
+    size_t samples_done = 0;
+    bool timestamp_set = false;
+    long waitTime = (timeoutUs == 0) ? SLEEP_1SEC : timeoutUs;
+
+    // Output pointer (void*, could be int16_t* or float*)
+    void *output_buf = buffs[0];
+
+    while (samples_done < numElems) {
+        // Consume any FIFO buffer leftovers first
```

Comment:
So after looking over this, you are changing things so the application calling readStream() can call with any amount of numElems.  correct?